Wrappers/Python/cil/optimisation/operators/Operator.py

```diff
@@ -678,9 +679,8 @@ def is_linear(self):
         return self.linear_flag
 
     def calculate_norm(self):
-        '''Returns the norm of the CompositionOperator, that is the product of the norms
-        of its operators.'''
-        norm = 1.
-        for operator in self.operators:
-            norm *= operator.norm()
-        return norm
+        '''Returns the norm of the CompositionOperator. Note that this gives a NotImplementedError if the CompositionOperator is not linear.'''
+        if self.is_linear():
+            return LinearOperator.calculate_norm(self)
+
+        return super().calculate_norm(self)

```

Comment:
again, there is nothing specific here for `CompositionOperator`, so there is no need to reimplement this function, as long as these lines are in `Operator`